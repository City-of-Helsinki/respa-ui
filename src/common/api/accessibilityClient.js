import { ApiClient } from './client';
import constants from '../../../app/constants/AppConstants';

export class AccessibilityApiClient {
  apiClient;

  systemId;

  constructor(baseUrl, systemId) {
    this.apiClient = new ApiClient(baseUrl, false);
    this.systemId = systemId;
  }

  getAccessibilityInformation(resourceId, viewPoints = [], locale = 'fi') {
    return Promise.all([
      this.getSentences(resourceId, locale),
      this.getShortcomings(resourceId, viewPoints, locale),
      // eslint-disable-next-line
    ]).then(([details, shortcomings]) => details.length > 0 || shortcomings.length > 0
      ? { details, shortcomings }
      : undefined);
  }

  getSentences(resourceId, locale = 'fi') {
    return this.apiClient
      .get(`targets/${this.systemId}/${resourceId}/sentences`)
      .then((res) => {
        return res.data
          .map((sentenceObject) => {
            return {
              sentenceGroup: sentenceObject.sentenceGroups.find(
                sentenceGroup => sentenceGroup.language === locale,
              ).value,
              sentences: sentenceObject.sentences
                .filter(sentence => sentence.language === locale)
                .map(sentence => sentence.value),
            };
          })
          .reduce((acc, sentenceObject) => {
            const index = acc.findIndex(
              accSentence => accSentence.sentenceGroup === sentenceObject.sentenceGroup,
            );

            if (index >= 0) {
              acc[index] = {
                ...acc[index],
                sentences: [
                  ...acc[index].sentences,
                  ...sentenceObject.sentences,
                ],
              };
              return acc;
            }

            return [...acc, sentenceObject];
          }, []);
      });
  }

  getShortcomings(resourceId, viewPoints = [], locale = 'fi') {
    // eslint-disable-next-line
    return Promise.all([
      this.apiClient.get(`targets/${this.systemId}/${resourceId}/shortages`),
      this.apiClient.get('viewpoints'),
    ]).then(([shortagesRes, viewpointsRes]) => {
      const idToNameMap = {};

      viewpointsRes.data.forEach((viewpoint) => {
        idToNameMap[viewpoint.viewpointId] = viewpoint.names.find(
          name => name.language === locale,
        ).value;
      });

      return shortagesRes.data
        .map(shortage => ({
          id: shortage.viewpointId,
          sentence: shortage.shortages.find(
            shortageSentence => shortageSentence.language === locale,
          ).value,
        }))
        .reduce((acc, shortage) => {
          const index = acc.findIndex(accItem => accItem.id === shortage.id);

          if (index >= 0) {
            acc[index] = {
              ...acc[index],
              sentences: [...acc[index].sentences, shortage.sentence],
            };
            return acc;
          }

          return [...acc, { id: shortage.id, sentences: [shortage.sentence] }];
        }, [])
        .filter(shortage => viewPoints.includes(shortage.id))
        .map(shortage => ({
          sentenceGroup: idToNameMap[shortage.id],
          sentences: shortage.sentences,
        }));
    });
  }
}

export default new AccessibilityApiClient(
  constants.ACCESSIBILITY_API_URL,
  constants.ACCESSIBILITY_API_SYSTEM_ID,
);
