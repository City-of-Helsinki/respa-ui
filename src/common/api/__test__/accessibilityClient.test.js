import axios from 'axios';

import { AccessibilityApiClient } from '../accessibilityClient';
import {
  mockSentencesResponseData,
  mockShortcomingsResponseData,
  mockViewpointsResponseData,
} from '../__fixtures__/accessibilityClientFixtures';

jest.mock('axios');

const apiUrl = 'https://test.test';
const systemId = '1234';
const resourceId = 1234;
const client = new AccessibilityApiClient(apiUrl, systemId);

describe('domain/common/api/accessibilityClient.js', () => {
  describe('getSentences', () => {
    beforeEach(() => {
      axios.request.mockResolvedValueOnce({
        data: mockSentencesResponseData,
      });
    });

    it('returns results in correct format', () => {
      expect(client.getSentences(resourceId)).resolves.toMatchSnapshot();
    });

    it('makes request to correct endpoint', () => {
      client.getSentences(resourceId);
      expect(axios.request.mock.calls[0][0].method).toBe('GET');
      expect(axios.request.mock.calls[0][0].url).toBe(
        `${apiUrl}/targets/${systemId}/${resourceId}/sentences/`,
      );
    });
  });

  describe('getShortcomings', () => {
    beforeEach(() => {
      axios.request
        .mockResolvedValueOnce({
          data: mockShortcomingsResponseData,
        })
        .mockResolvedValueOnce({ data: mockViewpointsResponseData });
    });

    it('returns results in correct format', () => {
      expect(client.getShortcomings(resourceId)).resolves.toMatchSnapshot();
    });

    it('makes requests to correct endpoints', () => {
      client.getShortcomings(resourceId);
      expect(axios.request.mock.calls[0][0].method).toBe('GET');
      expect(axios.request.mock.calls[0][0].url).toBe(
        `${apiUrl}/targets/${systemId}/${resourceId}/shortages/`,
      );
      expect(axios.request.mock.calls[1][0].method).toBe('GET');
      expect(axios.request.mock.calls[1][0].url).toBe(`${apiUrl}/viewpoints/`);
    });
  });
});
