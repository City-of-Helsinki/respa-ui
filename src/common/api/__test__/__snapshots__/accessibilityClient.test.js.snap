// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`domain/common/api/accessibilityClient.js getSentences returns results in correct format 1`] = `
Array [
  Object {
    "sentenceGroup": "Palvelutiski / vahtimestaripalvelut",
    "sentences": Array [
      "Kokoustila on rakennuksessa, jossa on palvelutiski / vahtimestaripalveluita. Ilmoittaudu ensin palvelutiskillä.",
      "Kokoustila sijaitsee samassa kerroksessa kuin sisäänkäynti.",
    ],
  },
  Object {
    "sentenceGroup": "Reitti kokoustilaan",
    "sentences": Array [
      "Matkalla kokoustilaan on lukittuja sähköisiä ovia, jotka avataan palvelutiskiltä tai ennakkoon hankittavalla kulunvalvonta-avaimella.",
      "Kulkureitillä kokoustilaan on ovi tai ovia, jotka aukeavat käsin helposti. Lasiovissa on kontrastimerkinnät.",
      "Kulkureitillä kokoustilaan on yli 2 cm korkeita kynnyksiä.",
    ],
  },
  Object {
    "sentenceGroup": "Kokoustila",
    "sentences": Array [
      "Kokoustilan ovi aukeaa kulunvalvonta-avaimella. Oviaukossa ei ole kynnystä.",
      "Kokoustilassa on niukasti tilaa liikkua pyörätuolin kanssa.",
      "Kokoustilassa on taustamelua, jonka lähde on ilmanvaihto. Kokolattiamatto vaimentaa ääntä.",
      "Kokoustilassa on istuimia. Kaikissa istuimissa on selkänoja.",
      "Kokoustilassa on pöytiä, joiden ääreen pääsee pyörätuolilla. Pöytien yhteydessä on pistorasioita kuten myös kokoustilan seinissä.",
      "Kokoustilan valaistus on tasainen ja häikäisemätön.",
    ],
  },
  Object {
    "sentenceGroup": "Esteetön wc",
    "sentences": Array [
      "Esteetön wc on samassa kerroksessa kuin kokoustila.",
      "Lähin miesten/naisten wc on samassa kerroksessa kuin kokoustila.",
    ],
  },
]
`;

exports[`domain/common/api/accessibilityClient.js getShortcomings returns results in correct format 1`] = `
Array [
  Object {
    "sentenceGroup": "Olen pyörätuolin käyttäjä",
    "sentences": Array [
      "Kulkureitillä kokoustilaan on yli 2 cm korkeita kynnyksiä.",
      "Kokoustilassa ei ole tarpeeksi tilaa liikkua pyörätuolilla.",
    ],
  },
  Object {
    "sentenceGroup": "Olen rollaattorin käyttäjä",
    "sentences": Array [
      "Kulkureitillä kokoustilaan on yli 2 cm korkeita kynnyksiä.",
      "Kokoustilassa ei ole tarpeeksi tilaa liikkua rollaattorilla.",
    ],
  },
  Object {
    "sentenceGroup": "Käytän kuulolaitetta",
    "sentences": Array [
      "Kokoustilassa ei ole kiinteää eikä siirrettävää induktiosilmukkaa.",
      "Kokoustilassa on häiritsevää taustamelua.",
    ],
  },
]
`;
